### PR TITLE
Replace CompletableFuture with ListenableFuture

### PR DESCRIPTION
**Pull request by dain**
The CompletableFuture interface is difficult to use correctly, and it is
easy to introduce subtle bugs that result in lost notifications. For example
CompletableFuture does not propagate cancel from wrapped futures to the
core future. Additionally, it is easy to accidentally schedule tasks on a system
wide fork join pool.

ListenableFuture on the other hand is easy to understand and reasonable, and
the transformations in Guava handle cancelation properly.

### CODE REVIEW

**Comment by martint** [comment]
Needs a commit message explaining why.

File: pom.xml [hunk 1]
```diff
@@ -46,7 +46,7 @@
         <air.maven.version>3.3.9</air.maven.version>
 
         <dep.antlr.version>4.6</dep.antlr.version>
-        <dep.airlift.version>0.142</dep.airlift.version>
+        <dep.airlift.version>0.143-SNAPSHOT</dep.airlift.version>
```
Comment (findepi) [inline pom.xml]: Can we base on a released version of airlift instead?

**Comment by ArturGajowy** [comment]
I'd strongly prefer we migrate to `CompletableFuture/CompletionStage` instead. This is because:
1. the API for `CompletableFuture` / `ComplestionStage` is much more high level and composable
2. `ListenableFuture` provokes constant usage of `isDone` and `get` because it extends Future. I'd prefer we had less of that boilerplate in favor of composing futures as it is possible with `CompletableFuture`
3. `CompletableFuture` is in the JDK

If we need to isolate users of the returned `Future`-s from the ability to complete the future, we can do so by exposing `CompletionStage` in the interface

**Comment by dain** [comment]
After working with `CompletableFuture/CompletionStage` for the past year it has become clear that the API is just too complex.  Typically when someone starts working on code that uses  `CompletableFuture` there is a long discussion about the exact semantics of the transform chain with regard to cancelation, and even things as exceptions (does thenApply trigger on exception?).  This is even true for those of us that have written massive amounts of code that uses `CompletableFuture`.

My comments apply to the Presto engine code, for the SPI we will continue to use CompletableFuture for now, but we will likely introduce an SPI interface to eliminate that usage also.  This will allow SPI implementors to choose the future completion framework that matches their code best (JDK, Guava, Netty Concurrent, etc).

File: pom.xml (hunk 1 above)
Comment (dain) [inline pom.xml]: Once the changes on the airlift side are reviewed, we can merge and release it.

File: presto-main/src/main/java/com/facebook/presto/server/TaskResource.java [hunk 2]
```diff
@@ -319,6 +324,20 @@ private static boolean shouldSummarize(UriInfo uriInfo)
         return uriInfo.getQueryParameters().containsKey("summarize");
     }
 
+    private static <T> ListenableFuture<T> addTimeout(ListenableFuture<T> future, Callable<T> onTimeout, Duration timeout, ScheduledExecutorService executorService)
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/server/TaskResource.java]: This should go as a separate commit as a util method in a future-related util class

File: presto-main/src/test/java/com/facebook/presto/execution/TestSqlTaskManager.java [hunk 3]
```diff
@@ -296,31 +296,31 @@ public ExchangeClient get(SystemMemoryUsageListener systemMemoryUsageListener)
         @Override
         public URI createQueryLocation(QueryId queryId)
         {
-            return URI.create("fake://query/" + queryId);
+            return URI.create("http://fake.invalid/query/" + queryId);
```
Comment (ArturGajowy) [inline presto-main/src/test/java/com/facebook/presto/execution/TestSqlTaskManager.java]: Unrelated changes

**Comment by dain** [comment]
Superseded by #7450
